README
Set_opacity - utility like transset, but more powerfull


ChangeLog
2006-02-09  XVilka Haos of System  <xvilka@gmail.com>
	* set_opacity.c: (main):
	Added option for selecting all windows
	* interface.c:
	Added function for selecting all windows
	* stack.c:
	Added STACK structure for easer windows id
	manipulating. 
	
	Many bugfixes. 

2005-09-17  XVilka Haos of System  <xvilka@gmail.com>
	* set_opacity.c: (main):
	Added support for long options
	Added verbose output
	* interface.c:
	Added select window by PID
	Added select window by name
	

stack.c
/***************************************************************************
                          stack.c  -  description
                             -------------------
    begin                : Wed Feb  9 14:03:54 EST 2006
    copyright            : (C) 2006-2010 by XVilka
    email                : xvilka@gmail.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include <stdlib.h>
#include "stack.h"

typedef struct STACKnode* link;
struct STACKnode
{
	Item item;
	link next;
};

static link head;

link NEW(Item item, link next)
{
	link x = malloc(sizeof *x);
	x->item = item;
	x->next = next;
	return x;
}

void STACKinit(int maxN)
{
	head = NULL;
}

int STACKempty(void)
{
	return head == NULL;
}

void STACKpush(Item item)
{
	head = NEW(item, head);
}

Item STACKpop(void)
{
	Item item = head->item;
	link t = head->next;
	free(head);
	head = t;
	return item;
}



stack.h
/***************************************************************************
                          stack.h  -  description
                             -------------------
    begin                : Wed Feb  9 14:03:54 EST 2006
    copyright            : (C) 2006-2010 by XVilka
    email                : xvilka@gmail.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

typedef int Item;
#define eq(A, B) (A == B)

void STACKinit(int maxN);
int STACKempty(void);
void STACKpush(Item item);
Item STACKpop(void);


interface.c
/***************************************************************************
                          interfaces.c  -  description
                             -------------------
    begin                : Wed Aug  17 14:03:54 EST 2005
    copyright            : (C) 2005-2010 by XVilka
    email                : xvilka@gmail.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include <X11/Xos.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/cursorfont.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>

#include "interface.h"
#include "stack.h"

unsigned window_debug = 0;
unsigned wid2pid_debug = 0;

void SetDebug(int debug)
{
	window_debug = debug;
	wid2pid_debug = debug;
}

Display *Open_Display(char *display_name)
{
	Display *dpy;

	dpy = XOpenDisplay(display_name);
	if (dpy == NULL)
	{
	    fprintf (stderr, "Unable to open display '%s'\n", XDisplayName (display_name));
	}
	return(dpy);
}

Window Get_Top_Window(Display *dpy, Window child)
{
	Window root, parent;
	Window *child_list;
	unsigned int num_children;

	if (!XQueryTree(dpy, child, &root, &parent, &child_list, &num_children))
  			fprintf(stderr, "Can't query window tree.\n");
  	if(parent == root) return child;
	while(parent != root)
	{
		child = parent;
		if (!XQueryTree(dpy, child, &root, &parent, &child_list, &num_children))
			fprintf(stderr, "Can't query window tree.\n");
	}
	return child;
}

Window Select_Window_by_Mouse(Display *dpy)
{
  int status;
  Cursor cursor;
  XEvent event;
  Window target_win = 0, root;
  int buttons = 0;
  unsigned screen = 0;

  root = RootWindow(dpy, screen);
  cursor = XCreateFontCursor(dpy, XC_crosshair);
  status = XGrabPointer(dpy, root, False, ButtonPressMask|ButtonReleaseMask, GrabModeSync,
			GrabModeAsync, root, cursor, CurrentTime);
  if (status != GrabSuccess)
  {
  		fprintf(stderr, "Can't grab the mouse!\n");
  		exit(1);
  }
  while ((target_win == 0) || (buttons != 0))
  {
    XAllowEvents(dpy, SyncPointer, CurrentTime);
    XWindowEvent(dpy, root, ButtonPressMask|ButtonReleaseMask, &event);
    switch (event.type)
    {
    	case ButtonPress:
      	if (target_win == 0)
      	{
      		target_win = event.xbutton.subwindow;
      		if (target_win == 0) target_win = root;
      	}
      	buttons++;
      	break;
    	case ButtonRelease:
      	if (buttons > 0)buttons--;
      	break;
    }
  }
  XUngrabPointer(dpy, CurrentTime);
  return(target_win);
}

Window Get_Window_by_Name(Display *dpy, Window top, char *name)
{
	Window *children, dummy;
	unsigned int nchildren;
	int i;
	Window w = 0;
	char *window_name;

	if (XFetchName(dpy, top, &window_name) && !strcmp(window_name, name)) return(top);
	if (!XQueryTree(dpy, top, &dummy, &dummy, &children, &nchildren)) return(0);
	for (i=0; i<nchildren; i++)
	{
		w = Get_Window_by_Name(dpy, children[i], name);
		if (w) break;
	}
	if (children) XFree ((char *)children);
	return(w);
}

Window Get_Window_by_PID(Display *dpy, pid_t pid)
{
        Window w = 0;
        Window root_return;
        Window parent_return;
        Window *children_return, *children_return1, *children_return2;
        unsigned int nchildren_return, nchildren_return1, nchildren_return2;
        unsigned i, j, k;

        if(window_debug) fprintf(stdout, "Check WINDOWID for pid=%d\n", pid);
        Status status = XQueryTree(dpy, DefaultRootWindow(dpy), &root_return,
        					&parent_return, &children_return, &nchildren_return);
        if(children_return)
        {
            if(window_debug) fprintf(stdout,"XQueryTree found %d child windows in root window\n",
            									nchildren_return);
            for (i = 0; i < nchildren_return; i++)
            {
                if (pid == WID2PID(dpy, children_return[i]))
                {
                   	w = children_return[i];
                    	break;
                }
                Status status1 = XQueryTree(dpy, children_return[i], &root_return,
                					&parent_return, &children_return1, &nchildren_return1);
                if(children_return1)
                {
                    if(window_debug) fprintf(stdout,"XQueryTree found %d child windows in window %lu(0x%lx)\n",
                                                            nchildren_return1, children_return[i], children_return[i]);
                    for (j = 0; j < nchildren_return1; j++)
                    {
                        if (pid == WID2PID(dpy, children_return1[j]))
                        {
                        	w = children_return1[j];
                        	break;
                        }
                        Status status2 = XQueryTree(dpy, children_return1[j], &root_return,
                        					&parent_return, &children_return2, &nchildren_return2);
                    		if(children_return2)
                    		{
                    				if(window_debug) fprintf(stdout,"XQueryTree found %d child windows in window %lu(0x%lx)\n",
                                                            nchildren_return2, children_return1[j], children_return1[j]);
                    				 for (k = 0; k < nchildren_return2; k++)
                    				 {
                    				 		if (pid == WID2PID(dpy, children_return2[k]))
                    				 		{
                    				 			w = children_return2[k];
                    				 			break;
                    				 		}
                    				 }
                    		 }
                    		 XFree(children_return2);
                    }
                    XFree(children_return1);
                }
                if(w) break;
            }
        }
        else
        {
        			fprintf(stdout, "No children return!\n");
        }
        XFree(children_return);
        if(!w){
        		if(window_debug) fprintf(stdout,"for pid=%d not found any window\n",pid);
        		return 0;
        }
        else {
            if(window_debug) fprintf(stdout, "for pid=%d found some window %lu(0x%lx).\n"
                                                " Interpret it as main application window\n",pid,w,w);
            return w;
        }
}

void Get_All_Windows(Display *dpy)
{
	Window w = 0;
        Window root_return;
        Window parent_return;
        Window *children_return, *children_return1, *children_return2;
        unsigned int nchildren_return, nchildren_return1, nchildren_return2;
        unsigned i, j, k;

        Status status = XQueryTree(dpy, DefaultRootWindow(dpy), &root_return,
        					&parent_return, &children_return, &nchildren_return);
        if(children_return)
        {
            STACKinit(100);
		if(window_debug) fprintf(stdout,"XQueryTree found %d child windows in root window\n",
            									nchildren_return);
            for (i = 0; i < nchildren_return; i++)
            {
                STACKpush(children_return[i]);
                Status status1 = XQueryTree(dpy, children_return[i], &root_return,
                					&parent_return, &children_return1, &nchildren_return1);
                if(children_return1)
                {
                    if(window_debug) fprintf(stdout,"XQueryTree found %d child windows in window %lu(0x%lx)\n",
                                                            nchildren_return1, children_return[i], children_return[i]);
                    for (j = 0; j < nchildren_return1; j++)
                    {
                        STACKpush(children_return1[j]);
                        Status status2 = XQueryTree(dpy, children_return1[j], &root_return,
                        					&parent_return, &children_return2, &nchildren_return2);
                    		if(children_return2)
                    		{
                    				if(window_debug) fprintf(stdout,"XQueryTree found %d child windows in window %lu(0x%lx)\n",
                                                            nchildren_return2, children_return1[j], children_return1[j]);
                    				 for (k = 0; k < nchildren_return2; k++)
                    				 {
                    				 	STACKpush(children_return2[k]);
                    				 }
                    		 }
                    		 XFree(children_return2);
                    }
                    XFree(children_return1);
                }
                if(w) break;
            }
        }
        else
        {
        			fprintf(stdout, "No children return!\n");
        }
        XFree(children_return);
}

pid_t WID2PID (Display *dpy, Window w)
{
    Atom atom,actual_type;
    char *atom_name;
    int actual_format;
    unsigned long nitems;
    unsigned long bytes_after;
    unsigned char *prop;
    int status;
    pid_t pid;

    if(wid2pid_debug) fprintf(stdout, "Check PID for wid=%d\n", w);

    atom = XInternAtom(dpy, "_NET_WM_PID", True);
    atom_name = XGetAtomName (dpy,atom);

    status = XGetWindowProperty(
        dpy, w, atom, 0, 1024,
        False, AnyPropertyType,
        &actual_type,
        &actual_format, &nitems,
        &bytes_after,
        &prop
    );

    if (status!=0)
    {
    	if(wid2pid_debug)  fprintf(stderr,"Cannot get _NET_WM_PID for window %lu(0x%lx)", w, w);
    	return -2;
    }
    if (! prop)
    {
    	if(wid2pid_debug)  fprintf(stderr,"window %lu(0x%lx): No properties\n", w, w);
    	return -3;
    }

    pid = prop[1] * 256;
    pid += prop[0];
    if(wid2pid_debug) fprintf(stdout,"pid of window %lu(0x%lx) = %d\n", w, w, pid);

    return pid;
}



interface.h
/***************************************************************************
                          interface.h  -  description
                             -------------------
    begin                : Wed Aug  17 14:03:54 EST 2005
    copyright            : (C) 2005-2010 by XVilka
    email                : xvilka@gmail.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#define ENV_SIZE 8192
#define OPAQUE	0xffffffff
#define OPACITY "_NET_WM_WINDOW_OPACITY"

void SetDebug(int debug);
Display *Open_Display(char *display_name);
Window Get_Top_Window(Display *dpy, Window child);
Window Select_Window_by_Mouse(Display *dpy);
Window Get_Window_by_Name(Display *dpy, Window top, char *name);
Window Get_Window_by_PID(Display *dpy, pid_t pid);
void Get_All_Windows(Display *dpy);
pid_t WID2PID (Display *dpy, Window w);


set_opacity.c
/***************************************************************************
                          set_opacity.c  -  description
                             -------------------
    begin                : Wed Aug  17 14:03:54 EST 2005
    copyright            : (C) 2005-2010 by XVilka
    email                : xvilka@gmail.com
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include <X11/Xlib.h>
#include <X11/Xatom.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>

#include "interface.h"
#include "stack.h"

const char* program_name;
Window target_win = 0;

void print_usage(FILE *stream, int exit_code)
{
	fprintf(stream, "Usage: %s options \n", program_name);
	fprintf(stream,
			" -h --help                           Display this usage information\n"
			" -v --verbose                        Print verbose messages\n"
			" -d --display [host]:display.screen  Select display and screen for usage\n"
			" -i --id PID                         Use pid for select window\n"
			" -m --mouse                          Use mouse for select window\n"
			" -a --all                            Select all windows\n"
			" -n --name NAME                      Use window name for selection\n"
			" -o --opacity 0 .. 1                 Window opacity\n"
			" -w --wid WID                        Use window id for selection\n");
	exit(exit_code);
}

int main(int argc, char *argv[])
{
  int next_option;
  const char *const short_options = "hvad:i:mn:w:o:";
  const struct option long_options[] = {
  		{ "help", 		0, NULL, 'h' },
  		{ "verbose",	0, NULL, 'v' },
		{ "all",		0, NULL, 'a' },
  		{ "display",	1, NULL, 'd' },
  		{ "id",			1, NULL, 'i' },
  		{ "mouse",		0, NULL, 'm' },
  		{ "name",		1, NULL, 'n' },
  		{ "wid",			1, NULL, 'w' },
  		{ "opacity",	1, NULL, 'o' },
  		{ NULL,			0, NULL, 0	 }
  };
  	int verbose = 0;
  	program_name = argv[0];

  	double vopacity;
  	unsigned int opacity;
  	char *display_name = NULL;
  	char *window_name = NULL;
  	pid_t pid = 0;
  	unsigned use_mouse = 1;
  	unsigned all = 0;
  	Window wid = 0;
  	unsigned char *data = NULL;
  	Atom actual;
  	int format;
  	unsigned long n, left;

  	do
	{
  		next_option = getopt_long(argc, argv, short_options, long_options, NULL);
  		switch (next_option)
  		{
  			case 'h':
  				print_usage(stdout, 0);
  			case 'v':
  				verbose = 1;
  				break;
  			case 'd':
  				display_name = optarg;
  				break;
  			case 'i':
  				pid = atoi(optarg);
  				use_mouse = 0;
  				break;
  			case 'm':
  				use_mouse = 1;
  				break;
  			case 'n':
  				window_name = optarg;
  				use_mouse = 0;
  				break;
  			case 'w':
  				wid = atoi(optarg);
  				use_mouse = 0;
  				break;
  			case 'o':
  				vopacity = atof(optarg);
  				opacity = 1;
  				break;
			case 'a':
				all = 1;
				pid = 0;
				wid = 0;
				window_name = NULL;
				break;
  			case '?':
  				print_usage(stderr, 1);
  			case -1:
  				break;
  			default:
  				abort();
  		}
  	}
  	while (next_option != -1);
  	if (verbose) SetDebug(1);

  	Display *dpy = Open_Display (display_name);
  	unsigned screen = DefaultScreen(dpy);

  	if (opacity)
  	{
  			opacity = (unsigned int) (vopacity * OPAQUE);
  	}
  	else
  	{
  			XGetWindowProperty(dpy, target_win, XInternAtom(dpy, OPACITY, False),
  						0L, 1L, False, XA_CARDINAL, &actual, &format, &n, &left, (unsigned char **) &data);
  			if (data != NULL)
  			{
  				memcpy (&opacity, data, sizeof (unsigned int));
  				XFree(( void *) data );
  				if (verbose) fprintf(stdout, "Found property: %g\n", (double) opacity / OPAQUE);
  			}
  			else opacity = OPAQUE;
  			if (opacity != OPAQUE) opacity = OPAQUE;
  			else opacity = 0xc0000000;
  	}
  	if (verbose) fprintf(stdout, "opacity 0x%x\n", opacity);
	if (!all)
	{
		if (use_mouse)
		{
			if(verbose) fprintf(stdout, "Using mouse for selection.\n");
			target_win = Select_Window_by_Mouse(dpy);
			if(verbose) fprintf(stdout, "Window id : %lu(0x%lx)\n", target_win, target_win);
		}
		else if (wid)
		{
			if(verbose) fprintf(stdout, "Using window id for selection.\n");
			target_win = wid;
		}
		else if (pid)
		{
			if(verbose) fprintf(stdout, "Using process id for selection.\n");
			target_win = Get_Window_by_PID(dpy, pid);
			if(target_win) target_win = Get_Top_Window(dpy, target_win);
			if(verbose) fprintf(stdout, "Window id : %lu(0x%lx)\n", target_win, target_win);
		}
		else if (window_name != NULL)
		{
			if(verbose) fprintf(stdout, "Using window name for selection.\n");
			target_win = Get_Window_by_Name(dpy, DefaultRootWindow(dpy), window_name);
			if(target_win) target_win = Get_Top_Window(dpy, target_win);
			if(verbose) fprintf(stdout, "Window id : %lu(0x%lx)\n", target_win, target_win);
		}
		if ((opacity == OPAQUE)&&(target_win != 0))
			XDeleteProperty (dpy, target_win, XInternAtom(dpy, OPACITY, False));
		else
			if (target_win != 0) XChangeProperty(dpy, target_win, XInternAtom(dpy, OPACITY, False),
				XA_CARDINAL, 32, PropModeReplace, (unsigned char *) &opacity, 1L);
    	}
	if (all)
	{
	  Get_All_Windows(dpy);
	  while(!STACKempty())
	  {
		 if (opacity == OPAQUE)
			 XDeleteProperty (dpy, STACKpop(), XInternAtom(dpy, OPACITY, False));
		 else
			 XChangeProperty(dpy, STACKpop(), XInternAtom(dpy, OPACITY, False),
				 XA_CARDINAL, 32, PropModeReplace, (unsigned char *) &opacity, 1L);
	  }
	}
	XSync(dpy, False);
	if(verbose) fprintf(stdout, "Set Property to %g\n", (double) opacity / OPAQUE);
	return EXIT_SUCCESS;
}
